keps/sig-cluster-lifecycle/kubeadm/4214-separate-super-user-kubeconfig/README.md

```diff
@@ -0,0 +1,724 @@
+kubeadm commands, such as `kubeadm upgrade`. In such conditions the `admin.conf`
+should be populated with credentials for a safe, temporary User that is bound
+to the `cluster-admin` ClusterRole.
+
+The super-user admin can then proceed to rotate the cluster certificate authority
+and eventually restore the `kubeadm:cluster-admins` ClusterRoleBinding.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Risk: Implementation complexity during in-place upgrade
+
+During `kubeadm upgrade apply` the new ClusterRoleBinding `kubeadm:cluster-admins`
+must be added. The switch to two separate files must be performed on all control plane
+nodes. This means that both `kubeadm upgrade apply` and `kubeadm upgrade node` must
+handle the migration of the `admin.conf` properly. The creation of the new
+`super-admin.conf` will be done only on the node where `kubeadm upgrade apply` is
+called. On later upgrades, one release after this feature is added, the certificate
+renewal logic of `kubeadm upgrade` must be aware that the `super-admin.conf` file could
+be missing and should not be rotated.
+
+Updating the `kube-system/kubeadm-certs` Secret contents where an encrypted
+`admin.conf` is stored will not be updated during upgrade.
+
+The mitigation here is detailed unit tests and e2e tests that ensure that
+the migration for in-place upgrades is handled properly.
+
+#### Risk: Implementation complexity during re-place upgrade
+
+Users or higher level tools that manage kubeadm re-place upgrades, by removing old
+control plane nodes and adding new control plane nodes, without calling
+`kubeadm upgrade apply/node` should handle this transition manually.
+The RBAC CLusterRoleBinding `kubeadm:cluster-admins` must be created before
+the upgrade has started. A new `admin.conf` that has the subject:
+`O = kubeadm:cluster-admins, CN = kubernetes-admin` must be uploaded in the
+`kube-system/kubeadm-certs` Secret and encrypted with the appropriate certificate key.
+Joining control plane nodes, must be able to download and decrypt the new `admin.conf`.
+
+Again, tests will be required to ensure that the `admin.conf` subject is migrated
+properly. The `super-admin.conf` file will not exist at all under such conditions,
+therefore the administrator can sign one manually by using
+the `kubeadm kubeconfig user --client-name=kubernetes-super-admin --org=system:masters`
+command.
+
+The same users or higher level tools can decide not to opt-in into this
+new behavior for existing clusters and continue using the `admin.conf` with
+`system:masters`. However, this means such clusters will drift away
+from the kubeadm security defaults.
```

Comment:
Does that mean we should support the legacy `admin.conf` always?

Reply:
the replace upgrades are a special case because they skip upgrade  apply/node. this means it's difficult for kubeadn upgrades to manage these users.

they can however decide to manage the kubeadm admin.conf the way they want, 1. update it to the new secure defaults, or 2. use the old way.

it's out of our control, but the change here will not break 2.

Reply:
sgtm!